### PR TITLE
Custom label

### LINKED ISSUE #697: Describe and implement rules on labels usage
https://kotlinlang.org/docs/reference/returns.html

add it to chapter 5

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
   Implemented rule, added tests and documentation.

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [ ] Added tests on fixers
* [x] Updated diktat-analysis.yml
* [x] Updated available-rules.md
* [x] closes #697

### CODE REVIEW

File: info/guide/guide-chapter-5.md [hunk 1]
```diff
@@ -146,3 +146,6 @@ GlobalScope.async {
 #### <a name="r5.2.5"></a> 5.2.5 Long lambdas should have explicit parameters
 The lambda without parameters shouldn't be too long.
 If a lambda is too long, it can confuse the user. Lambda without parameters should consist of 10 lines (non-empty and non-comment) in total.
+
+#### <a name="r5.2.6"></a> 5.2.6 Avoid using unnecessary, custom label
```
Comment (petertrr) [inline info/guide/guide-chapter-5.md:150]: Add this to TOC too

File: info/guide/guide-chapter-5.md [hunk 2]
```diff
@@ -146,3 +146,6 @@ GlobalScope.async {
 #### <a name="r5.2.5"></a> 5.2.5 Long lambdas should have explicit parameters
 The lambda without parameters shouldn't be too long.
 If a lambda is too long, it can confuse the user. Lambda without parameters should consist of 10 lines (non-empty and non-comment) in total.
+
+#### <a name="r5.2.6"></a> 5.2.6 Avoid using unnecessary, custom label 
+Expressions with unnecessary, custom labels generally increase complexity and worsen the maintainability of the code.
```
Comment (petertrr) [inline info/guide/guide-chapter-5.md:151]: Please describe exceptions and examples,

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter5/CustomLabelsTest.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                            q.forEach {
+                                return@qwe
+                            }
+                        }                
+                        q.forEachIndexed { index, i ->
+                            return@forEachIndexed
+                        }
+                        for(i: Int in q) {
+                            println(i)
+                            break@loop
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter5/CustomLabelsTest.kt:32]: Where is `loop@` defined? Will this compile?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    private val nestedParent = listOf(FOR, WHILE)
+    private lateinit var emitWarn: EmitType
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       emit: EmitType) {
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == LABEL_QUALIFIER && node.text !in labels) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt:32]: Does AST allow to distinguish label declaration and label usage?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import org.jetbrains.kotlin.psi.psiUtil.parents
+
+/**
+ * Rule that checks using custom label
+ */
+class CustomLabel(private val configRules: List<RulesConfig>) : Rule("custom-label") {
+    private var isFixMode: Boolean = false
+    private val forEachReference = listOf("forEach", "forEachIndexed")
+    private val labels = listOf("@loop", "@forEach", "@forEachIndexed")
+    private val nestedParent = listOf(FOR, WHILE)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt:23]: What about do-while?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       emit: EmitType) {
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == LABEL_QUALIFIER && node.text !in labels) {
+            val nestedCount = node.parents().count {
+                it.elementType in nestedParent ||
+                        (it.elementType == CALL_EXPRESSION && it.findChildByType(REFERENCE_EXPRESSION)?.text in forEachReference)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt:35]: I don't think this is necessary: for me using `forEach` inside raw loops seems like a bad idea anyway. And I don't see a test for this case.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt (hunk 4 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/CustomLabel.kt:32]: No , in AST they look same, that why I check `stopWords` near them

**Review by petertrr — APPROVED** [review]

**Pull request merged**